### PR TITLE
Update MerlinAU.sh Update to V1.0

### PR DESCRIPTION
**Pull request by ExtremeFiretop**
Update MerlinAU.sh Update to V1.0

Found a bug with my weighting system, it worked perfect for if you were on a beta, but if you were on production it broke the comparison lol As always, no winning.

Anyways, this time I promise I won over the problem. I removed padding though because our 32bit routers can't handle very large integers like 3004388006000 (32 bit system limit is: 2147483647) so instead it's now 300438860 without all the extra padding for comparison.

Now it seems to work correctly on betas and production, with the setting on and off.

### CODE REVIEW

**Comment by ExtremeFiretop** [comment]
Not ready for merge yet. Pending some more testing.

**Comment by ExtremeFiretop** [comment]
I think shes ready now! 🗡️ 
Except now you fixed the issue I fixed I see lol! I'll compare them to see.

How long would it take to add the unmount to the last step?

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub I think shes ready now! 🗡️
Except now you fixed the issue I fixed I see lol! I'll compare them to see.

How long would it take to add the unmount to the last step?

**Comment by Martinski4GitHub** [comment]
> Update MerlinAU.sh Update to V1.0
> 
> Found a bug with my weighting system, it worked perfect for if you were on a beta, but if you were on production it broke the comparison lol As always, no winning.
> 
> Anyways, this time I promise I won over the problem. I removed padding though because our 32bit routers can't handle very large integers like 3004388006000 (32 bit system limit is: 2147483647) so instead it's now 300438860 without all the extra padding for comparison.
> 
> Now it seems to work correctly on betas and production, with the setting on and off.
>
Yeah, found the same bug today while doing testing. I had to fix it right away in my local code so I could continue. I also added "**alpha**" version checking for those users who may go from "alpha" directly to production. It's likely very few who do that, but just in case I wanted to handle those scenarios.

**Comment by Martinski4GitHub** [comment]
> @Martinski4GitHub I think shes ready now! 🗡️ Except now you fixed the issue I fixed I see lol! I'll compare them to see.
> 
> How long would it take to add the unmount to the last step?
>
The code is almost ready, just double-checking and doing thorough testing to make sure only USB-attached mount points are unmounted. We don't want to accidentally unmount JFFS!!
Give me at least 45 minutes from now, I don't want to rush through it ("Slow & steady wins the race.").

**Comment by ExtremeFiretop** [comment]
> > @Martinski4GitHub I think shes ready now! 🗡️ Except now you fixed the issue I fixed I see lol! I'll compare them to see.
> > How long would it take to add the unmount to the last step?
> 
> The code is almost ready, just double-checking and doing thorough testing to make sure only USB-attached mount points are unmounted. We don't want to accidentally unmount JFFS!! Give me at least 45 minutes from now, I don't want to rush through it ("Slow & steady wins the race.").

That's okay, give me the same time lol! I need to actually review both our fixes.
What what I can tell so far, your using more complex regex usage while I'm using more complex conditional logic. Lol.

I could add alpha too, but that would grow my code more than yours since I'm using conditional logic I think...

**Comment by Martinski4GitHub** [comment]
> > > @Martinski4GitHub I think shes ready now! 🗡️ Except now you fixed the issue I fixed I see lol! I'll compare them to see.
> > > How long would it take to add the unmount to the last step?
> > 
> > 
> > The code is almost ready, just double-checking and doing thorough testing to make sure only USB-attached mount points are unmounted. We don't want to accidentally unmount JFFS!! Give me at least 45 minutes from now, I don't want to rush through it ("Slow & steady wins the race.").
> 
> That's okay, give me the same time lol! I need to actually review both our fixes. From what I can tell so far, your using more complex regex logic while I'm using more complex conditional logic. Lol.
> 
Yes, please double-check it to make sure it works with your testing as well. I did reduce the long numeric string so that, for example, now it's "3861204" instead of "386012004"

**Comment by ExtremeFiretop** [comment]
> > > > @Martinski4GitHub I think shes ready now! 🗡️ Except now you fixed the issue I fixed I see lol! I'll compare them to see.
> > > > How long would it take to add the unmount to the last step?
> > > 
> > > 
> > > The code is almost ready, just double-checking and doing thorough testing to make sure only USB-attached mount points are unmounted. We don't want to accidentally unmount JFFS!! Give me at least 45 minutes from now, I don't want to rush through it ("Slow & steady wins the race.").
> > 
> > 
> > That's okay, give me the same time lol! I need to actually review both our fixes. From what I can tell so far, your using more complex regex logic while I'm using more complex conditional logic. Lol.
> 
> Yes, please double-check it to make sure it works with your testing as well. I did reduce the long numeric string so that, for example, now it's "3861204" instead of "386012004"

Yeah I had reduced it as well. But my examples where like: 

> integers like 3004388006000 (32 bit system limit is: 2147483647) so instead it's now 300438860

So far so good, I'll probably merge yours in, and then do mine with my small other touch ups. So feel free to name it v0.99 ;)

**Comment by ExtremeFiretop** [comment]
Nvm.

Your version has a bug

**Comment by ExtremeFiretop** [comment]
Tbh I'm not sure how I feel about adding alpha, alpha to production? sounds sketchy, I was already on the fence about beta to production I only wanted to do production to production.

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

There I merged the 2 versions together, so mine has all your formatting fixes 👍 

I've still kept my logic because in my tests, going in debug mode with line: 2385 and flipping the switch via putty using: nvram set webs_state_flag=1 after starting the script, mine goes through for production and beta, your's only production.

**Comment by ExtremeFiretop** [comment]
Next I'll do a version check for BACKUPMON to be at least 1.44?

**Comment by ExtremeFiretop** [comment]
Done with the version check for BACKUPMON.
Only thing left is the unmount I guess?

**Comment by ExtremeFiretop** [comment]
Done for now :)
Might seems to work well, and I included all the changes from yours plus fixed the BACKUPMON stuff.

Whoot!

**Comment by Martinski4GitHub** [comment]
> Nvm.
> 
> Your version has a bug
> 
> This is yours doing 388.5.0 and it thinks it updated 
> 
> This is mine doing 388.5.0 and it pushes through:
> 
Yeah, I found the problem. It was caused by the maximum 32-bit signed integer limit when doing arithmetic operations with shell commands. It becomes a problem when the F/W "**branch**" version (**3004**.388.6.0) is included in the string; at that point, the number goes above the maximum integer. I fixed the problem by temporarily removing the "**branch**" version before doing arithmetic, and then putting it back into the resulting integer just before the value is returned. It's working well so far.

P.S. Sorry for the long delay. I had to run some errands with the wife that I had forgotten about.

**Comment by Martinski4GitHub** [comment]
> Tbh I'm not sure how I feel about adding alpha, alpha to production? sounds sketchy, I was already on the fence about beta to production I only wanted to do production to production.
>
I figured that if we're allowing "Beta to Production" we might as well allow "Alpha to Production" updates. It's essentially the same mechanism for checking & then updating.

**Comment by ExtremeFiretop** [comment]
> > Nvm.
> > Your version has a bug
> > This is yours doing 388.5.0 and it thinks it updated
> > This is mine doing 388.5.0 and it pushes through:
> 
> Yeah, I found the problem. It was caused by the maximum 32-bit signed integer limit when doing arithmetic operations with shell commands. It becomes a problem when the F/W "**branch**" version (**3004**.388.6.0) is included in the string; at that point, the number goes above the maximum integer. I fixed the problem by temporarily removing the "**branch**" version before doing arithmetic, and then putting it back into the resulting integer just before the value is returned. It's working well so far.
> 
> P.S. Sorry for the long delay. I had to run some errands with the wife that I had forgotten about.

No worries I got dinner in and some TV time after finishing my last test. And haha funny how we both ran into and resolved the same issue.

Only thing thing I need to review in this PR is if the version check for BACKUPMON is correct.

I'm not sure if the version.txt is the current version installed or the production updates it downloads to compare too. I think it's the latter in which case I did my logic wrong.

I know you added alpha, but I'm on the fence. I really only every intended this tool to be used on routers you want to "set and forget" and if your flashing alphas they can be buggy and I rather people handle them solo then blame the script.
 
Thoughts, opinions? You often have valid points I haven't considered so I'm curious why you wanted to add alpha?

**Comment by Martinski4GitHub** [comment]
> Done for now :) Mine seems to work well, and I included all the changes from yours plus fixed the BACKUPMON stuff.
> 
> Whoot!
>
I just submitted my changes to the PR now. It includes the fix plus the code to unmount the USB-attached drives.
I'll review your changes & double-check for any possible conflicts.

**Comment by Martinski4GitHub** [comment]
> 
> I know you added alpha, but I'm on the fence. I really only every intended this tool to be used on routers you want to "set and forget" and if your flashing alphas they can be buggy and I rather people handle them solo then blame the script.
> 
> Thoughts, opinions? You often have valid points I haven't considered so I'm curious why you wanted to add alpha?
>
I have no real strong opinions either way. I added it because I was making changes to that function and I thought we might as well have it just in case some users forgot to update from Alpha to Beta before updating to Production release. I've read posts of people in that situation asking if it was OK to update directly from Alpha to Production, and RMerlin responded that it was OK.

It's easy to remove it - one line change by removing the string from the **grep** command.

**Comment by ExtremeFiretop** [comment]
> > Done for now :) Mine seems to work well, and I included all the changes from yours plus fixed the BACKUPMON stuff.
> > Whoot!
> 
> I just submitted my changes to the PR now. It includes the fix plus the code to unmount the USB-attached drives. I'll review your changes & double-check for any possible conflicts.

The conflict will obviously be the one function we both fixed but I'll retest yours now, and if it works we can use the regex solution. I was willing to push mine through if I needed too but considering yours works for alpha too I guess users would like that more. My concern was a safety concern.

Everytime I had my issue it was from a beta firmware so it had me on the fence about it all.

**Comment by ExtremeFiretop** [comment]
> > I know you added alpha, but I'm on the fence. I really only every intended this tool to be used on routers you want to "set and forget" and if your flashing alphas they can be buggy and I rather people handle them solo then blame the script.
> > Thoughts, opinions? You often have valid points I haven't considered so I'm curious why you wanted to add alpha?
> 
> I have no real strong opinions either way. I added it because I was making changes to that function and I thought we might as well have it just in case some users forgot to update from Alpha to Beta before updating to Production release. I've read posts of people in that situation asking if it was OK to update directly from Alpha to Production, and RMerlin responded that it was OK.
> 
> It's easy to remove it - one line change by removing the string from the **grep** command.

If RMerlin says it's okay, I can argue.
I also sent you messages in the forums today to confirm I was able to upgrade from beta to production without issues as well so I guess it wasn't related to the beta after all.

**Comment by Martinski4GitHub** [comment]
> > 
> > I just submitted my changes to the PR now. It includes the fix plus the code to unmount the USB-attached drives. I'll review your changes & double-check for any possible conflicts.
> 
> The conflict will obviously be the one function we both fixed but I'll retest yours now, and if it works we can use the regex solution. I was willing to push mine through if I needed too but considering yours works for alpha too I guess users would like that more. My concern was a safety concern.
> 
> Everytime I had my issue it was from a beta firmware so it had me on the fence about it all.
>

Yeah, I'm reviewing your changes and keeping an eye out for conflicts. And yeah, I get your trepidation about allowing Alphas. Technically, as long as no changes have been made to the built-in F/W updater, our script should just work the same as when allowing Betas.

File: MerlinAU.sh [hunk 1]
```diff
@@ -1801,7 +1811,7 @@ _toggle_beta_updates_() {
 
     if [ "$currentSetting" = "ENABLED" ]; then
         printf "${REDct}*WARNING*:${NOct}\n"
-		printf "Disabling updates from beta F/Ws to release F/Ws may prevent access to the latest features and fixes.\n"
+		printf Disabling updates from beta to release firmware may limit access to new features and fixes.\n"
```
Comment (Martinski4GitHub) [inline MerlinAU.sh:1814]: The "**printf**" statement in line **1814** is missing the opening double quotes before the word "Disabling"

**Comment by ExtremeFiretop** [comment]
> > > I just submitted my changes to the PR now. It includes the fix plus the code to unmount the USB-attached drives. I'll review your changes & double-check for any possible conflicts.
> > 
> > 
> > The conflict will obviously be the one function we both fixed but I'll retest yours now, and if it works we can use the regex solution. I was willing to push mine through if I needed too but considering yours works for alpha too I guess users would like that more. My concern was a safety concern.
> > Everytime I had my issue it was from a beta firmware so it had me on the fence about it all.
> 
> Yeah, I'm reviewing your changes and keeping an eye out for conflicts. And yeah, I get your trepidation about allowing Alphas. Technically, as long as no changes have been made to the built-in F/W updater, our script should just work the same as when allowing Betas.

Just finished testing your new function and can confirm it works as expected with alphas and betas and production releases and my switch in the advanced settings also seems to work, so we now have 2 versions of the same function that work, yours however uses complex regex logic while mine uses complex conditional logic, yours is the better one for the fact it's tighten and more adaptive to alphas and betas, while mine is only for betas and while need more conditional checks for alphas growing it's size.

Lets go with yours. You also have the unmount which is highly valued before the flash.
I do however have good changes here, so I'll submit mine afterwards but yours wins for the conflict in that regard.

File: MerlinAU.sh (hunk 1 above)
Comment (ExtremeFiretop) [reply MerlinAU.sh:1814]: > The "**printf**" statement in line **1814** is missing the opening double quotes before the word "Disabling"

Good catch, fixing it now.

**Comment by Martinski4GitHub** [comment]
I'll resolve the conflicts UNLESS you're already doing it...

**Comment by ExtremeFiretop** [comment]
Resolved conflicts.

Ready for merge when you are. It cut my PR down in half after you merged yours LOL.

**Comment by ExtremeFiretop** [comment]
Ah I didn't review the version check for BACKUPMON

**Comment by Martinski4GitHub** [comment]
> Resolved conflicts.
> 
> Ready for merge when you are. It cut my PR down in half after you merged yours LOL.
> 
Ah, you beat me to it :>). Good thing I asked before jumping in.
Yep, ready as it can be.

**Comment by ExtremeFiretop** [comment]
> Ah I didn't review the version check for BACKUPMON

Yeah my version check for BM is garbage. Needs to be redone.

**Comment by Martinski4GitHub** [comment]
> Ah I didn't review the version check for BACKUPMON
>
You want to continue your review?? I'm not in a rush, so take your time.
If I understood correctly, we have until Wednesday, right?

**Comment by ExtremeFiretop** [comment]
> > Ah I didn't review the version check for BACKUPMON
> 
> You want to continue your review?? I'm not in a rush, so take your time. If I understood correctly, we have until Wednesday, right?

Your right we aren't in a rush, but I'm re-writing it now, I want your opinion on the route I'm taking, it won't be longer than 30 minutes, go grab a snack check on me and a few ;)

**Comment by Martinski4GitHub** [comment]
> > > Ah I didn't review the version check for BACKUPMON
> > 
> > 
> > You want to continue your review?? I'm not in a rush, so take your time. If I understood correctly, we have until Wednesday, right?
> 
> Your right we aren't in a rush, but I'm re-writing it now, I want your opinion on the route I'm taking, it won't be longer than 30 minutes, go grab a snack check on me and a few ;)
>
OK, just take your time. I'll take a break as I actually wanted to grab a snack. My wife & I went grocery shopping today and bought some bags of roasted almonds, cashews, peanuts, etc. I'll be back in 30 mins. Again, no rush... Slow but steady...

**Comment by ExtremeFiretop** [comment]
> > > > Ah I didn't review the version check for BACKUPMON
> > > 
> > > 
> > > You want to continue your review?? I'm not in a rush, so take your time. If I understood correctly, we have until Wednesday, right?
> > 
> > 
> > Your right we aren't in a rush, but I'm re-writing it now, I want your opinion on the route I'm taking, it won't be longer than 30 minutes, go grab a snack check on me and a few ;)
> 
> OK, just take your time. I'll take a break as I actually wanted to grab a snack. My wife & I went grocery shopping today and bought some bags of roasted almonds, cashews, peanuts, etc. I'll be back in 30 mins. Again, no rush... Slow but steady...

I'm done LOL.

Slow and steady is true, but I had already built the solution to parse from this .sh file, I just preferred to use the .txt file so submitted that instead. 

Then realize, if he uses the same update method we do, then that's a horrible way to check since that file is what it compares itself too for update checks and that file is always the latest lol.

Anyways all that to say, I had already build the code and save it on the side incase, and happy I did :) Just tested and everything seems functional to me, please review after your snack and let me know if I did any mistakes, if not.. This is officially production.

**Comment by ExtremeFiretop** [comment]
> > > > > Ah I didn't review the version check for BACKUPMON
> > > > 
> > > > 
> > > > You want to continue your review?? I'm not in a rush, so take your time. If I understood correctly, we have until Wednesday, right?
> > > 
> > > 
> > > Your right we aren't in a rush, but I'm re-writing it now, I want your opinion on the route I'm taking, it won't be longer than 30 minutes, go grab a snack check on me and a few ;)
> > 
> > 
> > OK, just take your time. I'll take a break as I actually wanted to grab a snack. My wife & I went grocery shopping today and bought some bags of roasted almonds, cashews, peanuts, etc. I'll be back in 30 mins. Again, no rush... Slow but steady...
> 
> I'm done LOL.
> 
> Slow and steady is true, but I had already built the solution to parse from this .sh file, I just preferred to use the .txt file so submitted that instead.
> 
> Then realize, if he uses the same update method we do, then that's a horrible way to check since that file is what it compares itself too for update checks and that file is always the latest lol.
> 
> Anyways all that to say, I had already build the code and save it on the side incase, and happy I did :) Just tested and everything seems functional to me, please review after your snack and let me know if I did any mistakes, if not.. This is officially production and ready as it'll ever be for me.

**Comment by ExtremeFiretop** [comment]
Done. Please review and merge to dev whenever your ready... After that. Production :D

**Comment by Martinski4GitHub** [comment]
> Done. Please review and merge to dev whenever your ready... After that. Production :D
>
Good to go, IMO.

**Comment by ExtremeFiretop** [comment]
> > Done. Please review and merge to dev whenever your ready... After that. Production :D
> 
> Good to go, IMO.

In that case, let me update the version number in dev and .txt file and push to Production? Any last things you want to add or change?

At this point I feel I've tested so much idk what else I could test lol.

**Comment by Martinski4GitHub** [comment]
> > > Done. Please review and merge to dev whenever your ready... After that. Production :D
> > 
> > 
> > Good to go, IMO.
> 
> In that case, let me update the version number in dev and .txt file and push to Production? Any last things you want to add or change?
> 
> At this point I feel I've tested so much idk what else I could test lol.
>
Just a question. Is the version string going to have 3 categories: "**major**.**minor**.**revision**"?
If so, should the 1st version be: "**1.0.0**"?
That's what I'm used to, and it makes it easy later on to compare versions like "**1.0.0**" vs "**1.0.2**" as you convert to numeric values. But I know other companies do it differently. There's really no standard. Either way, it's fine with me; I just wanted to know what our "standard" was going to be from the beginning going forward.

**Comment by ExtremeFiretop** [comment]
> > > > Done. Please review and merge to dev whenever your ready... After that. Production :D
> > > 
> > > 
> > > Good to go, IMO.
> > 
> > 
> > In that case, let me update the version number in dev and .txt file and push to Production? Any last things you want to add or change?
> > At this point I feel I've tested so much idk what else I could test lol.
> 
> Just a question. Is the version string going to have 3 categories: "**major**.**minor**.**revision**"? If so, should the 1st version be: "**1.0.0**"? That's what I'm used to, and it makes it easy later on to compare versions like "**1.0.0**" vs "**1.0.2**" as you convert to numeric values. But I know other companies do it differently. There's really no standard. Either way, it's fine with me; I just wanted to know what our "standard" was going to be from the beginning going forward.

That's a great question.

So I went with 1.0 only because that's what thelonelycoder asked for in the chat.

Quote: "The one thing i have now is the version bump to 1.0 if you feel comfortable."

If I look at other plugins on my signature, such as BACKUPMON, Viktor also have a 1 dot standard such as: (1.44). 
However other addons to have the 2 dot standard such as connmon v3.0.2 + scMerlin v2.4.0 + spdMerlin v4.4.5

So, really. Lets go with 1.0.0 because it's easier on us, and the lonely coder can clearly work with that standard.

**Comment by Martinski4GitHub** [comment]
> > 
> > Just a question. Is the version string going to have 3 categories: "**major**.**minor**.**revision**"? If so, should the 1st version be: "**1.0.0**"? That's what I'm used to, and it makes it easy later on to compare versions like "**1.0.0**" vs "**1.0.2**" as you convert to numeric values. But I know other companies do it differently. There's really no standard. Either way, it's fine with me; I just wanted to know what our "standard" was going to be from the beginning going forward.
> 
> That's a great question.
> 
> So I went with 1.0 only because that's what thelonelycoder asked for in the chat.
> 
> Quote: "The one thing i have now is the version bump to 1.0 if you feel comfortable."
> 
> If I look at other plugins on my signature, such as BACKUPMON, Viktor also have a 1 dot standard such as: (1.44). However other addons to have the 2 dot standard such as connmon v3.0.2 + scMerlin v2.4.0 + spdMerlin v4.4.5
>
YazFi & YazDHCP also have the 3-part version string. I think what thelonelycoder likely meant was to start with "1.0" as the 1st version & go forward from there, and *not* with "0.xx.yy" as we had been doing.

> 
> So, really. Lets go with 1.0.0 because it's easier on us, and the lonely coder can clearly work with that standard.
>
Yep, based on what I've seen with several add-ons, AMTM can work with the 3-part version.

**Comment by ExtremeFiretop** [comment]
> > > Just a question. Is the version string going to have 3 categories: "**major**.**minor**.**revision**"? If so, should the 1st version be: "**1.0.0**"? That's what I'm used to, and it makes it easy later on to compare versions like "**1.0.0**" vs "**1.0.2**" as you convert to numeric values. But I know other companies do it differently. There's really no standard. Either way, it's fine with me; I just wanted to know what our "standard" was going to be from the beginning going forward.
> > 
> > 
> > That's a great question.
> > So I went with 1.0 only because that's what thelonelycoder asked for in the chat.
> > Quote: "The one thing i have now is the version bump to 1.0 if you feel comfortable."
> > If I look at other plugins on my signature, such as BACKUPMON, Viktor also have a 1 dot standard such as: (1.44). However other addons to have the 2 dot standard such as connmon v3.0.2 + scMerlin v2.4.0 + spdMerlin v4.4.5
> 
> YazFi & YazDHCP also have the 3-part version string. I think what thelonelycoder likely meant was to start with "1.0" as the 1st version & go forward from there, and _not_ with "0.xx.yy" as we had been doing.
> 
> > So, really. Lets go with 1.0.0 because it's easier on us, and the lonely coder can clearly work with that standard.
> 
> Yep, based on what I've seen with several add-ons, AMTM can work with the 3-part version.

Done deal, moving forwards from today same rules apply we've generally been following I think, we use the 3 dot standard, and any changes are done through PRs in dev, if it's urgent, push it through right away and make the other a reviewer. 

Else, if it's not urgent it can steam in dev until the other gets the chance to review.
As always @Martinski4GitHub it was a pleasure, I actually really enjoyed the last few months and I feel like I've gotten a pretty good sense of who you are through all this.

Hope we continue to do great improvements over time! (Hopefully not too soon though! Haha!)

**Comment by Martinski4GitHub** [comment]
> 
> Done deal, moving forwards from today same rules apply we've generally been following I think, we use the 3 dot standard, and any changes are done through PRs in dev, if it's urgent, push it through right away and make the other a reviewer.
> 
Yes, we keep the same process to make changes: always via PRs submitted to the '**dev**' branch and have a 2nd pair of eyes review it to make sure & double-check. It never hurts. I would never want to push changes directly to the "**main**" branch unless it's something trivial like fixing spelling mistakes or adding comments to clarify some code. Even urgent fixes should be reviewed by someone else, just to verify. I know that when something looks urgent there's an impulse to fix it and release it right away, but I've seen situations made worse by releasing a bad fix, or a fix that breaks something else due to a lack of proper testing & validation. Managers tend to have that knee reaction, and I always advise for taking the time for proper review & testing. As the saying goes: "Measure twice and cut once."

>
> Else, if it's not urgent it can steam in dev until the other gets the chance to review. As always @Martinski4GitHub it was a pleasure, I actually really enjoyed the last few months and I feel like I've gotten a pretty good sense of who you are through all this.
>
Yes, it's been my pleasure to collaborate with you as well. I like how you handled yourself through this project. You don't take offense when I make suggestions that might conflict with your ideas, or when I make changes in your code that I believe make it better. I like your camaraderie and easy-going nature. I would not have continued working with you if every discussion had become an endless debate about who's right & who's wrong; if every little disagreement had become a dick-measuring contest.

I came in for the fun of coding and more learning, and that's what I got. I don't need any more "drama" in my life. There's enough of that at work with the usual "office politics" and the occasional "drama queen" that resurfaces when the pressure is on to deliver.

Anyway, I'm pleased with how things turned out & that I had the opportunity to meet you in the digital realm. You're one of the good guys, and I would be glad to call you my friend.

> 
> Hope we continue to do great improvements over time! (Hopefully not too soon though! Haha!)
>
Indeed.

**Comment by ExtremeFiretop** [comment]
> Done deal, moving forwards from today same rules apply we've generally been following I think, we use the 3 dot standard, and any changes are done through PRs in dev, if it's urgent, push it through right away and make the other a reviewer.


Absolutely, continuous improvement is the key, and it's even better when it's at a manageable pace. :) I like the fact that you keep a level head at the problems at hand, we will continue with the safety first mindset whenever possible going forwards.

Thank you for the kind words, I also like how you handled everything, you didn't take me asking questions to things personally, you really were a teacher and willing to walk me throw your thinking when I was clearly lost or confused or needed more info to understand what was going on, over time, I feel I've needed less hand-holding and I think that's just part of the journey working with someone as talented as you. Writing code with you is kinda like writing a book with you, at some point you get into a flow.

Looking forward to more projects where we can apply what we've learned and push the boundaries even further! 
It's been a rewarding journey so far, and I have no doubt it will continue to be. 
Here's to hoping our next challenges are just as engaging, but yeah, let's enjoy a bit of calm before the next storm! 

Keep in touch, and let's make sure to sync up for future endeavors. Happy sailing my friend! 
Cheers!

**Comment by ExtremeFiretop** [comment]
BTW ENJOY THOSE ROASTED ALMONDS, CASHEWS, PEANUTS, ETC.

**Comment by Martinski4GitHub** [comment]
> 
> Absolutely, continuous improvement is the key, and it's even better when it's at a manageable pace. :) I like the fact that you keep a level head at the problems at hand, we will continue with the safety first mindset whenever possible going forwards.
> 
> Thank you for the kind words, I also like how you handled everything, you didn't take me asking questions to things personally, you really were a teacher and willing to walk me throw your thinking when I was clearly lost or confused or needed more info to understand what was going on, over time, I feel I've needed less hand-holding and I think that's just part of the journey working with someone as talented as you. Writing code with you is kinda like writing a book with you, at some point you get into a flow.
> 
> Looking forward to more projects where we can apply what we've learned and push the boundaries even further! It's been a rewarding journey so far, and I have no doubt it will continue to be. Here's to hoping our next challenges are just as engaging, but yeah, let's enjoy a bit of calm before the next storm!
> 
> Keep in touch, and let's make sure to sync up for future endeavors. Happy sailing my friend! Cheers!
>
Godspeed!! & Take Care.

**Pull request merged**